File: apps/distill-electron/src/distill/logs.ts
import { openDistillDatabase } from "./db";
import { getBackgroundSyncStatus } from "./jobs";
import {
  DatasetExportTarget,
  ImportFailureEntry,
  ImportSourceSummary,
  LogEntry,
  LogEntryStatus,
  LogsPageData
} from "../shared/types";

type SyncJobRow = {
  id: number;
  status: string;
  last_error: string | null;
  payload_json: string;
  created_at: string;
  updated_at: string;
};

type ExportRow = {
  id: number;
  export_type: string;
  label_filter: string | null;
  output_path: string;
  record_count: number;
  metadata_json: string;
  created_at: string;
};

type SyncPayload = {
  reason?: string;
  startedAt?: string;
  finishedAt?: string;
  discoveredCaptures?: number;
  importedCaptures?: number;
  skippedCaptures?: number;
  failedCaptures?: number;
  summary?: string;
  sourceSummaries?: ImportSourceSummary[];
  failedEntries?: ImportFailureEntry[];
  outcome?: "completed" | "warning" | "failed";
};

type ExportPayload = {
  exportedAt?: string;
  dataset?: DatasetExportTarget;
};

function parseJson<T>(value: string, fallback: T): T {
  try {
    const parsed = JSON.parse(value) as T;
    return parsed && typeof parsed === "object" ? parsed : fallback;
  } catch {
    return fallback;
  }
}

function hasSyncWarnings(payload: Pick<SyncPayload, "failedCaptures" | "failedEntries">): boolean {
  return (payload.failedCaptures ?? 0) > 0 || (payload.failedEntries?.length ?? 0) > 0;
}

function normalizeSyncStatus(status: string, payload: SyncPayload): LogEntryStatus {
  if (status === "pending") {
    return "queued";
  }

  if (status === "running" || status === "warning" || status === "failed") {
    return status;
  }

  if (status === "completed") {
    if (payload.outcome) {
      return payload.outcome;
    }

    return hasSyncWarnings(payload) ? "warning" : "completed";
  }

  return "queued";
}

function summarizeSync(status: LogEntryStatus, payload: SyncPayload): string {
  if (payload.summary?.trim()) {
    return payload.summary.trim();
  }

  if (status === "queued") {
    return `Queued sync${payload.reason ? `: ${payload.reason}` : ""}`;
  }

  if (status === "running") {
    return `Sync running${payload.reason ? `: ${payload.reason}` : ""}`;
  }

  if (status === "warning") {
    return "Sync warnings";
  }

  if (status === "failed") {
    return "Sync failed";
  }

  return "Sync completed";
}

function stringifyRaw(value: Record<string, unknown>): string {
  return JSON.stringify(value, null, 2);
}

function normalizeExportDataset(value: string | null | undefined): DatasetExportTarget | undefined {
  if (value === "train" || value === "holdout") {
    return value;
  }

  return undefined;
}

function mapSyncJob(row: SyncJobRow): LogEntry {
  const payload = parseJson<SyncPayload>(row.payload_json, {});
  const status = normalizeSyncStatus(row.status, payload);
  const failedEntries = payload.failedEntries ?? [];
  const failedCaptures = payload.failedCaptures ?? 0;
  const level = status === "failed" ? "error" : "info";

  return {
    id: `sync-${row.id}`,
    kind: "sync",
    status,
    level,
    title: "Background sync",
    summary: summarizeSync(status, payload),
    createdAt: payload.startedAt ?? row.created_at,
    updatedAt: payload.finishedAt ?? row.updated_at,
    sourceLabel: payload.reason,
    metrics: {
      discoveredCaptures: payload.discoveredCaptures ?? 0,
      importedCaptures: payload.importedCaptures ?? 0,
      skippedCaptures: payload.skippedCaptures ?? 0,
      failedCaptures
    },
    details: {
      reason: payload.reason,
      sourceSummaries: payload.sourceSummaries ?? [],
      failedEntries
    },
    rawJson: stringifyRaw({
      jobId: row.id,
      status: row.status,
      lastError: row.last_error ?? undefined,
      ...payload
    })
  };
}

function mapExport(row: ExportRow): LogEntry {
  const payload = parseJson<ExportPayload>(row.metadata_json, {});
  const dataset = normalizeExportDataset(payload.dataset ?? row.label_filter);
  const datasetLabel = dataset ?? row.label_filter ?? "all";
  const recordLabel = row.record_count === 1 ? "record" : "records";

  return {
    id: `export-${row.id}`,
    kind: "export",
    status: "completed",
    level: "info",
    title: "Export",
    summary: `Exported ${row.record_count} ${datasetLabel} dataset ${recordLabel}`,
    createdAt: payload.exportedAt ?? row.created_at,
    updatedAt: payload.exportedAt ?? row.created_at,
    sourceLabel: datasetLabel,
    metrics: {
      recordCount: row.record_count
    },
    details: {
      dataset,
      outputPath: row.output_path
    },
    rawJson: stringifyRaw({
      exportId: row.id,
      exportType: row.export_type,
      dataset: datasetLabel,
      outputPath: row.output_path,
      recordCount: row.record_count,
      ...payload
    })
  };
}

function sortTime(entry: LogEntry): number {
  const timestamp = entry.updatedAt ?? entry.createdAt;
  if (!timestamp) {
    return 0;
  }

  const time = new Date(timestamp).getTime();
  return Number.isFinite(time) ? time : 0;
}

export function getLogsPageData(limit = 200): LogsPageData {
  const distillDb = openDistillDatabase();

  try {
    const syncJobs = distillDb.db
      .prepare(`
        SELECT id, status, last_error, payload_json, created_at, updated_at
        FROM jobs
        WHERE job_type = 'sync_sources'
        ORDER BY COALESCE(updated_at, created_at) DESC
        LIMIT ?
      `)
      .all(limit) as SyncJobRow[];

    const exports = distillDb.db
      .prepare(`
        SELECT id, export_type, label_filter, output_path, record_count, metadata_json, created_at
        FROM exports
        ORDER BY created_at DESC
        LIMIT ?
      `)
      .all(limit) as ExportRow[];

    const entries = [
      ...syncJobs.map(mapSyncJob),
      ...exports.map(mapExport)
    ]
      .sort((a, b) => sortTime(b) - sortTime(a))
      .slice(0, limit);

    return {
      entries,
      counts: {
        total: entries.length,
        errors: entries.filter((entry) => entry.level === "error").length,
        running: entries.filter((entry) => entry.status === "running").length
      },
      lastSyncStatus: syncJobs.length ? getBackgroundSyncStatus() : undefined
    };
  } finally {
    distillDb.close();
  }
}

File: apps/distill-desktop/src/data/logs.rs
use anyhow::Result;
use rusqlite::Connection;
use serde::{Deserialize, Serialize};

use crate::view_models::{LogEntryVm, LogFilter, LogsPageVm};

use super::{DesktopDataSource, matches_query, truncate_inline};

#[derive(Clone, Debug)]
struct LogRow {
    id: String,
    kind: String,
    title: String,
    subtitle: String,
    summary: String,
    status: String,
    level: String,
    metrics: String,
    raw_json: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
struct JobPayload {
    #[serde(default)]
    reason: Option<String>,
    #[serde(default)]
    summary: Option<String>,
    #[serde(default, rename = "startedAt")]
    started_at: Option<String>,
    #[serde(default, rename = "finishedAt")]
    finished_at: Option<String>,
    #[serde(default, rename = "discoveredCaptures")]
    discovered_captures: Option<i64>,
    #[serde(default, rename = "importedCaptures")]
    imported_captures: Option<i64>,
    #[serde(default, rename = "skippedCaptures")]
    skipped_captures: Option<i64>,
    #[serde(default, rename = "failedCaptures")]
    failed_captures: Option<i64>,
    #[serde(default)]
    outcome: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
struct ExportPayload {
    #[serde(default, rename = "exportedAt")]
    exported_at: Option<String>,
    #[serde(default)]
    dataset: Option<String>,
}

impl DesktopDataSource {
    pub fn load_logs(
        &self,
        filter: LogFilter,
        query: &str,
        expanded_log_id: Option<&str>,
    ) -> Result<LogsPageVm> {
        if !self.database_exists() {
            return Ok(LogsPageVm {
                entries: Vec::new(),
                summary_total_text: "0 entries".to_string(),
                summary_error_text: "0 errors".to_string(),
                summary_sync_text: "idle".to_string(),
                empty_title: "No logs yet".to_string(),
                empty_message: match self.source_mode() {
                    crate::config::SourceMode::RustOwned => {
                        "Sync jobs and exports from the Rust-owned Distill store will appear here when present."
                            .to_string()
                    }
                    crate::config::SourceMode::ElectronCompatReadOnly => {
                        "Open a compatible Distill Electron database to inspect sync and export history."
                            .to_string()
                    }
                },
            });
        }

        let conn = self.open_read_only()?;
        let mut rows = self.load_log_rows(&conn)?;
        rows.sort_by(|left, right| right.subtitle.cmp(&left.subtitle));
        let total_count = rows.len();
        let latest_sync_label = rows
            .iter()
            .find(|row| row.kind == "sync")
            .map(|row| normalize_summary_sync_label(row))
            .unwrap_or_else(|| "idle".to_string());
        let error_count = rows.iter().filter(|row| row.level == "error").count();

        rows.retain(|row| {
            matches_log_filter(row, filter)
                && matches_query(
                    query,
                    &[&row.title, &row.summary, &row.metrics, &row.raw_json],
                )
        });

        let expanded_id = expanded_log_id
            .filter(|candidate| rows.iter().any(|row| row.id == *candidate))
            .map(ToOwned::to_owned);

        let entries = rows
            .into_iter()
            .map(|row| LogEntryVm {
                expanded: expanded_id.as_deref() == Some(row.id.as_str()),
                id: row.id,
                title: row.title,
                subtitle: row.subtitle,
                summary: row.summary,
                status: row.status,
                level: row.level,
                metrics: row.metrics,
                raw_json: row.raw_json,
            })
            .collect::<Vec<_>>();

        let (empty_title, empty_message) = if entries.is_empty() {
            if query.trim().is_empty() && matches!(filter, LogFilter::All) {
                (
                    "No logs yet".to_string(),
                    match self.source_mode() {
                        crate::config::SourceMode::RustOwned => {
                            "Sync and export activity will show up here once the Rust-owned Distill store has operational history."
                                .to_string()
                        }
                        crate::config::SourceMode::ElectronCompatReadOnly => {
                            "Sync and export activity will show up here once Distill Electron has operational history."
                                .to_string()
                        }
                    },
                )
            } else {
                (
                    "No matching logs".to_string(),
                    "Adjust the log search or filters to widen the current result set."
                        .to_string(),
                )
            }
        } else {
            (String::new(), String::new())
        };

        Ok(LogsPageVm {
            entries,
            summary_total_text: format!("{total_count} entries"),
            summary_error_text: format!("{error_count} errors"),
            summary_sync_text: latest_sync_label,
            empty_title,
            empty_message,
        })
    }

    fn load_log_rows(&self, conn: &Connection) -> Result<Vec<LogRow>> {
        let mut rows = Vec::new();

        let mut job_stmt = conn.prepare(
            r#"
            SELECT id, status, last_error, payload_json, created_at, updated_at
            FROM jobs
            WHERE job_type = 'sync_sources'
            ORDER BY COALESCE(updated_at, created_at) DESC
            "#,
        )?;
        let mut job_rows = job_stmt.query([])?;
        while let Some(row) = job_rows.next()? {
            let id: i64 = row.get(0)?;
            let status: String = row.get(1)?;
            let last_error: Option<String> = row.get(2)?;
            let payload_json: String = row.get(3)?;
            let created_at: String = row.get(4)?;
            let updated_at: Option<String> = row.get(5)?;
            let payload = serde_json::from_str::<JobPayload>(&payload_json).unwrap_or_default();
            let normalized = normalize_sync_status(&status, &payload);
            let metrics = format!(
                "{} found · {} imported · {} skipped · {} failed",
                payload.discovered_captures.unwrap_or(0),
                payload.imported_captures.unwrap_or(0),
                payload.skipped_captures.unwrap_or(0),
                payload.failed_captures.unwrap_or(0)
            );
            rows.push(LogRow {
                id: format!("sync-{id}"),
                kind: "sync".to_string(),
                title: "Background sync".to_string(),
                subtitle: updated_at.unwrap_or(created_at),
                summary: payload.summary.clone().unwrap_or_else(|| {
                    payload
                        .reason
                        .clone()
                        .unwrap_or_else(|| "Sync activity".to_string())
                }),
                status: normalized.clone(),
                level: if normalized == "failed" {
                    "error".to_string()
                } else {
                    "info".to_string()
                },
                metrics,
                raw_json: serde_json::to_string_pretty(&serde_json::json!({
                    "jobId": id,
                    "status": status,
                    "lastError": last_error,
                    "payload": payload,
                }))
                .unwrap_or_else(|_| payload_json.clone()),
            });
        }

        let mut export_stmt = conn.prepare(
            r#"
            SELECT id, export_type, label_filter, output_path, record_count, metadata_json, created_at
            FROM exports
            ORDER BY created_at DESC
            "#,
        )?;
        let mut export_rows = export_stmt.query([])?;
        while let Some(row) = export_rows.next()? {
            let id: i64 = row.get(0)?;
            let export_type: String = row.get(1)?;
            let label_filter: Option<String> = row.get(2)?;
            let output_path: String = row.get(3)?;
            let record_count: i64 = row.get(4)?;
            let metadata_json: String = row.get(5)?;
            let created_at: String = row.get(6)?;
            let payload = serde_json::from_str::<ExportPayload>(&metadata_json).unwrap_or_default();
            let dataset = payload
                .dataset
                .clone()
                .or(label_filter.clone())
                .unwrap_or_else(|| "all".to_string());
            rows.push(LogRow {
                id: format!("export-{id}"),
                kind: "export".to_string(),
                title: "Export".to_string(),
                subtitle: created_at,
                summary: format!("Exported {record_count} {dataset} records"),
                status: "completed".to_string(),
                level: "info".to_string(),
                metrics: format!(
                    "type={export_type} · output={}",
                    truncate_inline(&output_path, 48)
                ),
                raw_json: serde_json::to_string_pretty(&serde_json::json!({
                    "exportId": id,
                    "exportType": export_type,
                    "dataset": dataset,
                    "outputPath": output_path,
                    "recordCount": record_count,
                    "payload": payload,
                }))
                .unwrap_or_else(|_| metadata_json.clone()),
            });
        }

        Ok(rows)
    }
}

fn normalize_sync_status(status: &str, payload: &JobPayload) -> String {
    match status {
        "pending" => "queued".to_string(),
        "running" | "warning" | "failed" => status.to_string(),
        "completed" => {
            if let Some(outcome) = payload.outcome.as_deref() {
                outcome.to_string()
            } else if payload.failed_captures.unwrap_or(0) > 0 {
                "warning".to_string()
            } else {
                "completed".to_string()
            }
        }
        other => other.to_string(),
    }
}

fn normalize_summary_sync_label(row: &LogRow) -> String {
    match row.status.as_str() {
        "warning" => "sync warnings".to_string(),
        "failed" => "sync failed".to_string(),
        "running" => "syncing...".to_string(),
        "queued" => "sync queued".to_string(),
        "completed" => format!("synced {}", row.subtitle),
        _ => row.status.clone(),
    }
}

fn matches_log_filter(row: &LogRow, filter: LogFilter) -> bool {
    match filter {
        LogFilter::All => true,
        LogFilter::Sync => row.kind == "sync",
        LogFilter::Export => row.kind == "export",
        LogFilter::Errors => row.level == "error",
    }
}
